### PR TITLE
Add types and client for Resolution

### PR DESCRIPTION
**Pull request by abayer**
# Changes

Part of #4710

In order to get remote resolution functionality to beta, we're moving it from a separate repository/release to be part of Pipeline. This is the first PR in a sequence moving the code from the Resolution repository into Pipeline.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
action required: Remote resolver `resource` field has changed to `params`.
```

### CODE REVIEW

**Comment by abayer** [comment]
/assign @vdemeester @dibyom 
cc @chitrangpatel

**Comment by abayer** [comment]
/hold

Putting this on hold until we've got v0.38.0 cut, 'cos I'd really like to have all of the relevant commits in one release together.

File: pkg/apis/resolution/v1alpha1/resolution_request_types.go
Comment (pritidesai) [inline pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: I was sharing the remote resource spec with a team and came across two different syntaxes. It seems we are implementing it using `params`:

```
taskRef:
    resolver: git
    resource:
    - name: url
      value: https://github.com/tekton/catalog.git
    - name: path
      value: /task/golang-fuzz/0.1/golang-fuzz.yaml
    - name: branch
      value: main
```

The [proposal](https://github.com/tektoncd/community/blob/main/teps/0060-remote-resource-resolution.md#proposal) in the Remote Resolution TEP has listed it as:

```
taskRef:
  resolver: git
  resource:
    repository_url: https://github.com/tektoncd/catalog.git
    branch: main
    path: /task/golang-fuzz/0.1/golang-fuzz.yaml
```

Do we need to update the proposal to reflect `params`? or is it documented somewhere else?

Comment (abayer) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: Good question! I didn't do the initial implementation, but while I like the `repository_url` etc syntax better, the `- name: url` syntax is simpler to implement, since we don't need to use a `runtime.RawExtension` here for arbitrary YAML. 

So I'd say the proposal should be updated, yeah.

Comment (vdemeester) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: @abayer to play devil's advocate, we should aim for the API to be the easiest to use (for end-user) than the easiest to implement 😝

Comment (abayer) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: Fair point. =) I spent some time this weekend thinking about this - in practice, it probably is easier to go with a simple `map[string]string` than I'd been thinking, given that what we've got now is already just a slice of string name/value pairs anyway. Then, if we hit a situation down the road where we have a need for more complex resolver parameters, we could change that to something like `unstructured.Unstructured` without breaking API compatibility.

Comment (abayer) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: ...though it is worth mentioning that changing the syntax will break compatibility with the existing syntax, but that's alpha syntax and probably not at all widely used right now anyway.

Comment (vdemeester) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: > ...though it is worth mentioning that changing the syntax will break compatibility with the existing syntax, but that's alpha syntax and probably not at all widely used right now anyway.

yeah that's the only reason that was "against" changing, but this is behind the alpha feature-flag for that sole reason, so if we properly communicate, I think that is fine.

Comment (abayer) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: 👍 Lemme do that!

**Comment by abayer** [comment]
@pritidesai @vdemeester I've changed the `resource:` syntax to be 
```
resource:
  foo: bar
```
instead of
```
resource:
  - name: foo
     value: bar
```

**Review by vdemeester — APPROVED** [review]
/cc @tektoncd/core-maintainers

File: pkg/apis/pipeline/v1beta1/resolver_types.go
Comment (abayer) [reply pkg/apis/pipeline/v1beta1/resolver_types.go:36]: Very good question! I am not a fan of `resource` here - it's not really clear what it means in the first place. I feel like just `params` would work, though since `params` is used in a bunch of places with different syntax than here, I'm not sure if that's ideal...

**Comment by abayer** [comment]
/hold

Putting on hold again until we nail down whether we keep the `resource` field name or not.

File: pkg/apis/pipeline/v1beta1/resolver_types.go
Comment (vdemeester) [reply pkg/apis/pipeline/v1beta1/resolver_types.go:36]: In the TEP, we used `resource` ? I'd rather go with `params` or `values` than `resource` indeed.

**Comment by abayer** [comment]
I've added a commit changing `resource:` to `params:` - if it seems good to everyone, I'll squash it, and open a corresponding PR to update the TEP.

**Comment by lbernick** [comment]
OK I'm sorry I don't really have a better suggestion here, but `params` might be confusing since the syntax was changed to be non-params-like

**Comment by abayer** [comment]
@lbernick Yeah, I don't like `params` either, but can't think of anything better. `values` doesn't feel right...maybe `args`? Though it's not the same as a container/step's `args`...

**Comment by abayer** [comment]
You know what? I'm just going to go with `paramsMap` for now - yeah, it's still got `params` in it, but I feel like being a different string, with `map` explicitly in the string, is good enough.

**Review by lbernick — CHANGES_REQUESTED** [review]
Sorry to push back on this but I think we should discuss naming at API WG if reviewers for this PR are able to attend-- I'll try to brainstorm some ideas before then but I think getting naming correct is important

**Comment by abayer** [comment]
No worries! =) I'm not in love with any option at this point, though I do prefer `paramsMap` to either `params` or `resource`, hence me updating the PR.

**Comment by abayer** [comment]
Another option could be `resolverParams` - which stutters a bit being under the `resolver` field. Again, I'm open to literally any suggestion.

**Comment by abayer** [comment]
So after discussion with @lbernick and her discovery of https://github.com/tektoncd/pipeline/pull/4502#discussion_r803199842, I've come to the conclusion that we should stick with the subobject syntax, with the field renamed to `params`. As discussed in that comment, this is more consistent with Kubernetes in general, and it also will make it much easier to add support for array/object parameters in the future if we so desire. That will entail a change to `ResolutionRequestSpec.Parameters` to not be a `map[string]string` as well, and an update to the TEP.

cc @pritidesai @vdemeester

File: pkg/apis/resolution/v1alpha1/resolution_request_validation.go
Comment (dibyom) [inline pkg/apis/resolution/v1alpha1/resolution_request_validation.go:22]: not for this PR but at some point we should get rid of the "common" package and just have the helper function in better named packages

**Review by dibyom — APPROVED** [review]

File: pkg/apis/resolution/v1alpha1/resolution_request_validation.go
Comment (abayer) [reply pkg/apis/resolution/v1alpha1/resolution_request_validation.go:22]: Good idea - I'll do that in the PR where I move that package in.

**Review by lbernick — APPROVED** [review]
Thanks Andrew! LGTM but I'll let Vincent or Priti give the final lgtm just to make sure they are also on board with this syntax.

Please don't forget the action required release note!

**Comment by vdemeester** [comment]
> So after discussion with @lbernick and her discovery of [#4502 (comment)](https://github.com/tektoncd/pipeline/pull/4502#discussion_r803199842), I've come to the conclusion that we should stick with the subobject syntax, with the field renamed to `params`. As discussed in that comment, this is more consistent with Kubernetes in general, and it also will make it much easier to add support for array/object parameters in the future if we so desire. That will entail a change to `ResolutionRequestSpec.Parameters` to not be a `map[string]string` as well, and an update to the TEP.
> 
> cc @pritidesai @vdemeester

So, I agree that, if we stick with `params` field for this, we should be consistent with the shape of `params` in the other parts of the API. I disagree that we should be necessarily consistent with Kubernetes on this. We should *aim* for the best UX for our users not because Kubernetes does it — and we are trying to not be too much "tied up" with Kubernetes as well. If we aim for end-users to author `Pipeline`, `Task`, … we should aim for making the API tailored for them, and not necessarily follow what *idiomatic* Kubernetes API does/looks like.

The `params` syntax is quite verbose. If we use the `bundle` for comparison :

```yaml
# from
taskRef:
  bundle: quay.io/vdemeest/foo:bar
  name: baz
# to
taskRef:
  resolver: bundle
  params:
  - name: image
    value: quay.io/vdemeest/foo:bar
  - name: name
    value: baz
  - name: type
    value: Task
```

From a user perspective, an "author", I have the feeling that the following feels better.

```yaml
taskRef:
  resolver: bundle
  params:
    image: quay.io/vdemeest/foo:bar
    name: baz
    type: Task
```

That *might* be one of the reason the [TEP](https://github.com/tektoncd/community/blob/main/teps/0060-remote-resource-resolution.md) uses `resource` instead, to not have to be consistent with `params`.

On "additionnal" point here, the resulting `ResolutionRequest` type will have a different `params` syntax as the one specified in `resolver`. Do we want to change this as well ?

```yaml
apiVersion: resolution.tekton.dev/v1alpha1
kind: ResolutionRequest
# […]
spec:
  params:
    serviceAccount: default
    bundle: gcr.io/tekton-releases/catalog/upstream/golang-build:0.1
    name: golang-build
    kind: task
EOF
```

In a gist, I am not a huge fan of using `params` and the `task.params` syntax, and I would rather stick to the proposed TEP syntax 😅 ; mainly because, from a user/author perspective, it's easier to write 😇. *Sorry for the back and forth 😓*

**Comment by abayer** [comment]
> On "additionnal" point here, the resulting `ResolutionRequest` type will have a different `params` syntax as the one specified in `resolver`. Do we want to change this as well ?

Eventually, yes, but I think that will involve moving `ArrayOrString` out of `pkg/apis/pipeline` so we can use it in `pkg/apis/resolution` as well. Since that will be a decent amount of work, and `ResolutionRequest`s aren't user-authored anyway, I thought that work could wait til later.

**Comment by vdemeester** [comment]
> it's "use the same admittedly subpar/overly verbose syntax for `resolver.[params|resource]` that we do elsewhere or use a different syntax unique to this one field to accomplish the same end result".

I agree with that 💯 % 😝. But *if* we didn't name it `params`, but something else, maybe `spec` 😈. Then we would have something completely "new" and "unique", and more closely to the `spec` you tend to see in Kubernetes. For example `volumeClaimTemplate`, etc.. I agree it's a very thin line between parameters and the "spec" of the resolver though 😝.

**Comment by abayer** [comment]
> I agree with that 💯 % 😝. But _if_ we didn't name it `params`, but something else, maybe `spec` 😈. Then we would have something completely "new" and "unique", and more closely to the `spec` you tend to see in Kubernetes. For example `volumeClaimTemplate`, etc.. I agree it's a very thin line between parameters and the "spec" of the resolver though 😝.

Right - so I think that's a legitimate argument for `ResolutionRequest`'s `Params` field, since that's in a different CRD. But within the context of a `pipelineRef` or `taskRef`, I think it would be a mistake to have a different syntax for `.resolver.[params|resource]` than we do elsewhere in the same `Pipeline`/`PipelineRun`/`TaskRun` CRD.

**Comment by abayer** [comment]
In general, I'd argue that we should only have one syntax in Pipeline for arbitrarily-named key/value pairs, and since we already have one (`params` and `results` etc using `- name: ...`), we should keep using that one syntax in all cases.

File: pkg/apis/resolution/v1alpha1/resolution_request_types.go
Comment (jerop) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go]: the reason why it was implemented as such was because of K8s convention: "Lists of named subobjects preferred over maps" - as discussed in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps and https://github.com/kubernetes/kubernetes/issues/2004

and we listed in https://github.com/tektoncd/community/blob/main/design-principles.md#api-conventions that Tekton API should comply with the K8s API conventions

Comment (jerop) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go]: haha just saw the comments below

**Comment by abayer** [comment]
FYI, I'm reworking `[PipelineRef|TaskRef].ResolverRef.Params` and `ResolutionRequestSpec.Params` to be `[]v1beta1.Param` now, since it feels a little silly to not do so, with validation, replacements, etc supported.

**Comment by abayer** [comment]
For what it's worth, in the process of switching over to using `[]v1beta1.Param` everywhere, I realized another reason for using the `Param` syntax (and therefore using `v1beta1.Param` behind the scenes - it makes param/result substitution simpler, and that's something I think users will end up wanting. By using `v1beta1.Param`, we just have to add the appropriate calls to the standard replacement functions for `v1beta1.Param` without having to write new substitution logic specifically for a new type.

**Comment by abayer** [comment]
I've just switched `ResolutionRequest`'s spec to keep the map for the moment, so that we can have the CRD be the same as it is in Resolution currently, and we'll follow that up with a deprecation and switch to `.params` syntax there later. Note that the `pipelineRef` and `taskRef` syntax is still changing from `resource` as a map of strings to `params` as a slice of `v1beta1.Param`.

**Comment by abayer** [comment]
So this still needs a squash, but I've avoided doing so until there's buy-in from @vdemeester. =)

I deliberately decided to _not_ change the `ResolutionRequest` syntax for now, because there's already the v1alpha1 version in Resolution. Once this chain lands, I'll deprecate v1alpha1 and replace it with v1beta1 using `[]Param` rather than `map[string]string`, but I didn't want to break compatibility for the existing CRD. We still need to have the v1alpha1 definition here so that we can add conversion to it, so that's why it's a bit messy. =)

File: pkg/apis/resolution/v1alpha1/resolution_request_defaults.go
Comment (abayer) [reply pkg/apis/resolution/v1alpha1/resolution_request_defaults.go:22]: We have - I tried to avoid having `pkg/apis/resolution` depend on anything from `pkg/apis/pipeline`, which may have been overkill.

File: pkg/apis/resolution/v1alpha1/resolution_request_types.go
Comment (abayer) [reply pkg/apis/resolution/v1alpha1/resolution_request_types.go:64]: ...and yeah, I reverted back to the old form for `ResolutionRequest.Spec` so as not to break compatibility. See https://github.com/tektoncd/pipeline/pull/5200#issuecomment-1219463898 =)

File: pkg/apis/pipeline/v1beta1/pipeline_types_test.go
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipeline_types_test.go:344]: `ArrayOrString` was renamed to `ParamValue` in https://github.com/tektoncd/pipeline/pull/5304 -- we need to rename it in this PR as well

**Review by jerop — COMMENTED** [review]
`ArrayOrString` was renamed to `ParamValue` in https://github.com/tektoncd/pipeline/pull/5304 -- we need to rename it in this PR as well

**Comment by abayer** [comment]
> `ArrayOrString` was renamed to `ParamValue` in #5304 -- we need to rename it in this PR as well

Done!

**Pull request merged**
